File: advent_of_code/2021/1/solution.py
"""
Solution for AoC 2021 Day 1
https://adventofcode.com/2021/day/1
"""
from typing import List


def count_increases_with_sliding_window(nums: List[int], window_size: int = 1) -> int:
    """
    Complexity analysis
    Time: O(n)
    Space: O(1)
    """
    count = 0

    for i in range(0, len(nums) - window_size):
        if nums[i + window_size] > nums[i]:
            count += 1

    return count


def main():
    with open("input.txt") as handle:
        nums = [int(line) for line in handle]

    print(f"Step 1: {count_increases_with_sliding_window(nums)}")
    print(f"Step 2: {count_increases_with_sliding_window(nums, sliding_window_size=3)}")


if __name__ == "__main__":
    main()

File: advent_of_code/2021/1/solution.go
package main

// Solution for AoC 2021 Day 1
// https://adventofcode.com/2021/day/1

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func ReadInts(r io.Reader) ([]int, error) {
	scanner := bufio.NewScanner(r)
	scanner.Split(bufio.ScanWords)

	var ints []int

	for scanner.Scan() {
		i, err := strconv.Atoi(scanner.Text())
		if err != nil {
			return ints, err
		}
		ints = append(ints, i)
	}

	return ints, scanner.Err()
}

func CountIncreasesWithSlidingWindow(ints []int, windowSize int) int {
	/*
	   Complexity analysis:
	   Time: O(n)
	   Space: O(1)
	*/
	count := 0

	for i := 0; i < len(ints)-windowSize; i++ {
		if ints[i+windowSize] > ints[i] {
			count++
		}
	}

	return count
}

func main() {
	handle, err := os.Open("input.txt")
	check(err)

	defer handle.Close()

	nums, err := ReadInts(handle)
	check(err)

	fmt.Printf("Part 1: %d\n", CountIncreasesWithSlidingWindow(nums, 1))
	fmt.Printf("Part 2: %d\n", CountIncreasesWithSlidingWindow(nums, 3))
}
